crates/devolutions-pedm/src/utils.rs

```diff
@@ -0,0 +1,161 @@
+use devolutions_pedm_shared::policy::{Hash, User};
+use digest::Update;
+use sha1::Sha1;
+use sha2::{Digest, Sha256};
+use std::collections::HashMap;
+use std::fs;
+use std::path::Path;
+use tracing::info;
+use win_api_wrappers::raw::Win32::Foundation::{GENERIC_ALL, GENERIC_READ};
+use win_api_wrappers::raw::Win32::Security::Authorization::SE_FILE_OBJECT;
+use win_api_wrappers::raw::Win32::Security::{WinLocalSystemSid, OBJECT_INHERIT_ACE};
+use win_api_wrappers::win::{
+    create_directory, set_named_security_info, Account, Ace, AceType, Acl, CommandLine, InheritableAcl,
+    InheritableAclKind, SecurityAttributes, SecurityDescriptor, Sid,
+};
+use win_api_wrappers::{
+    raw::Win32::System::Threading::PROCESS_CREATION_FLAGS,
+    win::{create_process_as_user, ProcessInformation, StartupInfo, Token},
+};
+
+use anyhow::Result;
+
+pub fn start_process(
+    token: &Token,
+    executable_path: Option<&Path>,
+    command_line: Option<&CommandLine>,
+    inherit_handles: bool,
+    creation_flags: PROCESS_CREATION_FLAGS,
+    environment: Option<&HashMap<String, String>>,
+    current_directory: Option<&Path>,
+    startup_info: &mut StartupInfo,
+) -> Result<ProcessInformation> {
+    let token = token.duplicate_impersonation()?;
+    let account = token.sid_and_attributes()?.sid.account(None)?;
+
+    info!(
+        ?executable_path,
+        ?command_line,
+        account.account_name,
+        "Starting process"
+    );
+
+    let mut process_information = None;
+    token.impersonate(|| {
+        process_information = Some(create_process_as_user(
+            Some(&token),
+            executable_path,
+            command_line,
+            None,
+            None,
+            inherit_handles,
+            creation_flags,
+            environment,
+            current_directory,
+            startup_info,
+        )?);
+
+        Ok(())
+    })?;
+
+    Ok(process_information.unwrap())
+}
+
+#[derive(Default)]
+pub struct MultiHasher {
+    sha1: Sha1,
+    sha256: Sha256,
+}
+
+impl MultiHasher {
```

Comment:
Why do we actually need multi hash? is it required for WinAPI calls or for the other reason?

Reply:
Not really. It's more of a convenience feature. If the user wants to add a hash check they can select either sha256 or sha1. There's sha1 since it's popular for certificate fingerprints. This is just a helper that computes all the supported hashes as one.